lib/src/license_detection/new_tokenizer.dart

```diff
@@ -0,0 +1,256 @@
+import 'dart:collection';
+
+import 'package:meta/meta.dart';
+import 'package:source_span/source_span.dart';
+import 'package:string_scanner/string_scanner.dart';
+
+@sealed
+class Token {
+  /// Normalized form of the text in [fileSpan].
+  final String value;
+
+  /// Denotes the token position.
+  final int tokenID;
+
+  /// Zero based line number of token.
+  final int line;
+
+  /// SourceSpan of the token.
+  final SourceSpan fileSpan;
+
+  Token(this.value, this.tokenID, this.line, this.fileSpan);
+
+  Token.fromSpan(this.fileSpan, this.tokenID)
+      : value = _normalizeWord(fileSpan.text.toLowerCase()),
+        line = fileSpan.start.line;
+}
+
+List<Token> tokenizer(String text) {
```

Comment:
Can you write a formal definition of what the tokens this returns are?

What is a token? Not how does a text get tokenized, but what is the end result that comes out.
What kind of tokens are created? In the final output is there tokens for new lines? Is there token for whitespace? Are punctuations embedded in tokens?

Another option is to describe this with a sequence of examples of what results from:
 * `hello world`
 * `hello world!`
 * `hello world !`
 * `hello world\n!`
 * `hello !world`
 * `word 42`
 * `word42`
 * `word.42`
 * `one word. Other word`
 * ` 'quotes' `
 * ` 'quotes" `
 * ` [brackets] `
 * ` [ brackets ] `
 * ` { brackets } `
 * ` ( brackets ) `
 * ` word & word`
 * ` word &word`
 * ` word&word`
 * ` word-word`
 * ` word -word`
 * ` word - word`
 * ` word- word`
 * `one TWO\n * Tree fouR \n * FIVE six!\n`

It's not obvious from the code how these should be tokenized, and it's hard to see if the code does the right things when there is no definition of what it is supposed to do. Maybe these could be test cases.. and maybe the documentation here only needs a few examples :D

But the more I read here, the more I'm thinking we need to first figure out what the result of tokenization and normalization is supposed to be.